src/utils/embedded/embedded.provider.tsx

```diff
@@ -259,37 +264,122 @@ export function EmbeddedProvider({ children }: EmbeddedProviderProps) {
+    };
 
-    const {
-      shareHash: recoveryBackupShareHash,
-      sharePublicKey: recoveryBackupSharePublicKey
-    } = await WalletUtils.generateShareHashAndPublicKey(recoveryBackupShare);
+    const hasRecoveryShareLocally = await hasStoredRecoveryShare();
+    if (hasRecoveryShareLocally) {
+      const storedRecovery = await retrieveStoredRecoveryShare();
+      if (storedRecovery) {
+        recoveryFileData = storedRecovery;
+      }
+    }
 
-    const { recoveryFileServerSignature, wallet: updatedWallet } =
-      await WalletService.registerRecoveryShare({
-        walletId,
-        recoveryAuthShare,
-        recoveryBackupShareHash,
-        recoveryBackupSharePublicKey
-      });
+    if (
+      !recoveryFileData.recoveryBackupShare &&
+      !recoveryFileData.recoveryFileServerSignature
+    ) {
+      const { recoveryAuthShare, recoveryBackupShare } =
+        await WalletUtils.generateWalletRecoveryShares(jwk);
 
-    updateCurrentWallet((currentWallet) => ({
-      ...currentWallet,
-      ...updatedWallet
-    }));
+      const {
+        shareHash: recoveryBackupShareHash,
+        sharePublicKey: recoveryBackupSharePublicKey
+      } = await WalletUtils.generateShareHashAndPublicKey(recoveryBackupShare);
 
-    downloadRecoveryFile(walletAddress, {
-      walletId,
-      recoveryBackupShare,
-      recoveryFileServerSignature
-    });
+      const { recoveryFileServerSignature, wallet: updatedWallet } =
+        await WalletService.registerRecoveryShare({
+          walletId,
+          recoveryAuthShare,
+          recoveryBackupShareHash,
+          recoveryBackupSharePublicKey
+        });
+
+      updateCurrentWallet((currentWallet) => ({
+        ...currentWallet,
+        ...updatedWallet
+      }));
+
+      recoveryFileData = {
+        ...recoveryFileData,
+        recoveryBackupShare,
+        recoveryFileServerSignature
+      };
+
+      // Store encrypted recovery share in local storage if feature flag is enabled
+      if (EMBEDDED_FEATURE_FLAGS.STORE_RECOVERY_SHARES) {
+        try {
+          // Create the recovery file data
+          const recoveryData = {
+            version: "1",
+            ...recoveryFileData
+          } as RecoveryJSON;
+
+          await WalletUtils.storeEncryptedRecoveryShare(
+            walletId,
+            recoveryData,
+            jwk
+          );
+          console.log("Recovery share stored successfully");
+        } catch (error) {
+          console.error("Failed to store recovery share:", error);
+        }
+      }
+    }
+
+    // Download the recovery file for the user
+    downloadRecoveryFile(walletAddress, recoveryFileData);
 
     // TODO: Make sure we use `freeDecryptedWallet` all over the place in the new code for Embedded:
     freeDecryptedWallet(jwk);
   }, [walletId, walletAddress]);
 
+  // Check if a wallet has a stored recovery share
+  const hasStoredRecoveryShare = useCallback(async (): Promise<boolean> => {
```

Comment:
I would remove `hasStoredRecoveryShare` and `retrieveStoredRecoveryShare` from the provider, as these functions are not used outside of it.

Reply:
Sure, will remove it from the provider.